### PR TITLE
Use the base class in Bio.SeqIO.Interface to check the number of sequences while writing

### PR DESCRIPTION
**Pull request by mdehoon**
This pull request uses the `Writer` base class in Bio.SeqIO.Interface to check the number of sequences while writing a sequence file.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

**Comment by peterjc** [comment]
Why not just update the existing classes like adding ``mincount`` and ``maxcount`` to the ``SequentialSequenceWriter`` method ``write_records``?

**Comment by mdehoon** [comment]
@peterjc  We can do that too, but in any case we need only one base class; having both `SequenceWriter` and `SequentialSequenceWriter` is confusing. But if we modify either of those, we'll need deprecation warnings and whatnot, so it's easier to just make a new class. Also,  `SequenceWriter` is a bit if a misnomer, as it writes sequence records, not sequences.

**Comment by peterjc** [comment]
I think the new class is so similar that it could replace ``SequentialSequenceWriter `` (and if there is no use case left for it, deprecate ``SequenceWriter``).

That way the only API changes look like adding a few optional arguments?

**Comment by mdehoon** [comment]
But then `SequentialSequenceWriter` wouldn't just be for sequential file formats.

**Comment by peterjc** [comment]
OK, replace ``SequenceWriter`` and assuming no use case left for it, deprecate ``SequentialSequenceWriter``?

**Comment by chris-rands** [comment]
would it make sense to make `Writer` an Abstract Base Class?

**Comment by peterjc** [comment]
@chris-rands Maybe, but I don't know without more reading. The answer probably applies to all the existing classes in ``Bio/SeqIO/Interfaces.py``

**Comment by mdehoon** [comment]
@chris-rands  Either `write_record` or `write_file` should be implemented in the derived class, and declared abstract in the abstract base class. However, to then instantiate a member of the base class, we need to implement *both* `write_record` and `write_file` in the derived class, which doesn't make sense in most cases.

**Comment by mdehoon** [comment]
@peterjc 

> OK, replace SequenceWriter and assuming no use case left for it, deprecate SequentialSequenceWriter?

Yes we can do that. So then, in the current pull request I would remove `SequenceWriter`, and rename `Writer` to `SequenceWriter` ?

**Comment by peterjc** [comment]
That sounds right - hopefully the class is in the same place in the file, so the diff will be minimal?

**Comment by chris-rands** [comment]
@mdehoon Yes, i see, this design does not really fit since one can't have optional either/or`@abstractmethod`s. Maybe `SequenceIterator` could be an `ABC` with `__next__ ` as an `@abstractmethod` though? In fact, they are several other base classes in the codebase that could be ABCs.

@peterjc  to clarify, the advantage of ABCs i think is that they forbid creating instances of the base class and forbid sub-classing without defining the required `@abstracmethod`s. This is more strict than raising `NotImplemented` since that only fails when the method is called. With Python 3, the syntax is simple

Anyway, I don't want to interfere with what you're doing here, later (when I have time and after this is merged), I'll open a PR to try this, if you think it's worth pursuing?

**Comment by peterjc** [comment]
Sounds good Chris. This stuff was written over ten years ago and pre-dates ABCs in Python.

**Comment by mdehoon** [comment]
@chris-rands  
> Maybe SequenceIterator could be an ABC with __next__ as an @abstractmethod though?
Possibly, but what would you be iterating over for the interlaced formats?
Perhaps it's easiest if we move ahead with this PR now, and then you can have another look at how to use an ABC?

File: Bio/SeqIO/Interfaces.py [hunk 1]
```diff
@@ -141,25 +147,91 @@ def clean(self, text):
...
+        records - A list or iterator returning SeqRecord objects
         """
-        # Note when implementing this, your writer class should NOT close the
-        # file at the end, but the calling code should.
-        raise NotImplementedError("This object should be subclassed")
-        #####################################################
-        # You SHOULD subclass this                          #
-        #####################################################
+        ####################################################
+        # You SHOULD implement this method in the subclass #
```
Comment (peterjc) [inline Bio/SeqIO/Interfaces.py]: Remove this comment, or at least rephrase. When would you want to implement this?

File: Bio/SeqIO/Interfaces.py [hunk 2]
```diff
@@ -141,25 +147,91 @@ def clean(self, text):
...
+        ####################################################
+
+    def write_record(self, record):
+        """Write a single record to the output file.
+
+        record - a SeqRecord object
+        """
+        raise NotImplementedError("This method should be implemented")
+        ####################################################
+        # You SHOULD implement this method in the subclass #
```
Comment (peterjc) [inline Bio/SeqIO/Interfaces.py]: SHOULD -> MUST

**Comment by mdehoon** [comment]
> Squash and merge for a simpler history?

To be honest, I don't know what that means. I guess I can just hit the big button with "Squash and merge" in red on it? But I don't see any other button, so I wouldn't know what the alternative would be.

**Comment by peterjc** [comment]
It should go green once the automated tests have finished, but yes.

There is a little drop down arrow on the right hand edge of the button, so you can change it from Squash-and-merge (you get a single commit on the master branch with all the changes from the pull request) to Rebase-and-merge (if the PR had 10 separate commits, you get 10 separate commits on the master).

Squash-and-merge is a safe default, especially if there was a lot of back-and-forth with ideas tried and reverted.

The rebase-and-merge is particularly useful when the commits are logically separated and you want to keep that in the master branch history, or the PR contains commits authored by different people.

See also ``git merge --squash ...`` and ``git rebase ...`` at the command line.

**Comment by chris-rands** [comment]
I can't see where/how is `SequenceIterator` actually used? Is it just a legacy code?

**Comment by peterjc** [comment]
It looks like ``SequenceIterator`` is now unused - I think that means all the parsers ended up being implemented as generator functions, rather than iterator classes. Perhaps it can be deprecated...

**Comment by chris-rands** [comment]
since `SequenceIterator` is unused, we can definitely safely make it an ABC ;)

But seriously, I think there's quite a bit of biopython code like this, maybe a cleanout with vulture would help https://github.com/jendrikseipp/vulture

**Comment by peterjc** [comment]
Vulture sounds worth a try - go for it!

**Pull request merged**